aospy/automate.py

```diff
@@ -0,0 +1,220 @@
+    """
+    try:
+        input = raw_input
+    except NameError:
+        import builtins
+        input = builtins.input
+    return input
+
+
+def _user_verify():
+    input = _input_func_py2_py3()
+    if not input("Perform these computations? [y/n] ").lower()[0] == 'y':
+        raise IOError('\n', 'Execution cancelled by user.')
+
+
+class CalcSuite(object):
+    """Generate suite of Calc objects based on inputted specifications."""
+
+    _CORE_SPEC_NAMES = ('projects',
+                        'models',
+                        'runs')
+
+    _AUX_SPEC_NAMES = ('variables',
+                       'date_ranges',
+                       'regions',
+                       'input_time_intervals',
+                       'input_time_datatypes',
+                       'input_time_offsets',
+                       'input_vertical_datatypes',
+                       'output_time_intervals',
+                       'output_time_regional_reductions',
+                       'output_vertical_reductions')
+
+    _NAMES_SUITE_TO_CALC_MAPPING = {
+        'projects': 'proj',
+        'models': 'model',
+        'runs': 'run',
+        'variables': 'var',
+        'date_ranges': 'date_range',
+        'regions': 'region',
+        'input_time_intervals': 'intvl_in',
+        'input_time_datatypes': 'dtype_in_time',
+        'input_time_offsets': 'time_offset',
+        'input_vertical_datatypes': 'dtype_in_vert',
+        'output_time_intervals': 'intvl_out',
+        'output_time_regional_reductions': 'dtype_out_time',
+        'output_vertical_reductions': 'dtype_out_vert',
+    }
+
+    def __init__(self, calc_suite_specs, obj_lib):
+        """Turn all inputs into aospy-ready objects."""
+        self._specs_in = calc_suite_specs
+        self._obj_lib = obj_lib
+
+        self._projects = [self._obj_lib.projects[project_name]
+                          for project_name in self._specs_in['projects']]
+
+    def _permute_core_specs(self):
+        obj_trees = []
+        for project in self._projects:
+            models = _get_mult_objs_by_names(self._specs_in['models'], project,
+                                             'models', 'default_models')
+            for model in models:
+                runs = _get_mult_objs_by_names(self._specs_in['runs'], model,
+                                               'runs', 'default_runs')
+                for run in runs:
+                    # Use Calc's attr names as keys.
+                    obj_trees.append(dict(proj=project, model=model, run=run))
+        return obj_trees
+
+    def _permute_aux_specs(self):
+        # Drop the "core" specifications, which are handled separately.
+        specs = self._specs_in.copy()
+        [specs.pop(core) for core in self._CORE_SPEC_NAMES]
+        # TODO: don't just pull from first project.
+        regions = _get_mult_objs_by_names(specs['regions'], self._projects[0],
```

Comment:
Yes that's a good point...I need to think about this

A related issue (although I don't see an immediate use-case) is if the same name was used by Region objects in different projects but with different region definitions.  In that case this logic would incorrectly use the region definition for the first project in all cases.